modules/javafx.media/src/main/native/gstreamer/plugins/mfwrapper/mfgstbuffer.cpp

```diff
@@ -0,0 +1,262 @@
+
+    return S_OK;
+}
+
+// IUnknown
+HRESULT CMFGSTBuffer::QueryInterface(REFIID riid, void **ppvObject)
+{
+    if (!ppvObject)
+    {
+        return E_POINTER;
+    }
+    else if (riid == IID_IUnknown)
+    {
+        (*ppvObject) = static_cast<IUnknown *>(static_cast<IMFMediaBuffer *>(this));
+    }
+    else if (riid == IID_IMFMediaBuffer)
+    {
+        (*ppvObject) = static_cast<IMFMediaBuffer *>(this);
+    }
+    else
+    {
+        (*ppvObject) = NULL;
+        return E_NOINTERFACE;
+    }
+    AddRef();
+    return S_OK;
+}
+
+ULONG CMFGSTBuffer::AddRef()
+{
+    return InterlockedIncrement(&m_ulRefCount);
+}
+
+ULONG CMFGSTBuffer::Release()
+{
+    ULONG uCount = InterlockedDecrement(&m_ulRefCount);
+    if (uCount == 0)
+    {
+        delete this;
+    }
+    return uCount;
+}
+
+// GStreamer interface
+HRESULT CMFGSTBuffer::GetGstBuffer(GstBuffer **ppBuffer)
+{
+    if (ppBuffer == NULL)
+        return E_INVALIDARG;
+
+    // If we do not have GStreamer buffer or it is still locked
+    // return E_UNEXPECTED. Such condition should not happen, but
+    // just in case we need to check for it.
+    if (m_pGstBuffer == NULL || m_bUnmapGstBuffer)
+        return E_UNEXPECTED;
+
+    (*ppBuffer) = m_pGstBuffer;
+
+    m_pGstBuffer = NULL;
+
+    return S_OK;
+}
+
+HRESULT CMFGSTBuffer::SetCallbackData(sCallbackData *pCallbackData)
+{
+    if (pCallbackData == NULL)
+        ZeroMemory(&m_CallbackData, sizeof(sCallbackData));
+    else
+        m_CallbackData = (*pCallbackData);
+
+    return S_OK;
+}
+
+HRESULT CMFGSTBuffer::SetGetGstBufferCallback(void (*function)(GstBuffer **ppBuffer,
+            long lSize, sCallbackData *pCallbackData))
+{
+    if (function == NULL)
+        return E_INVALIDARG;
+
+    GetGstBufferCallback = function;
+
+    return S_OK;
+}
+
+HRESULT CMFGSTBuffer::AllocateOrGetBuffer(BYTE **ppbBuffer)
+{
+    if (ppbBuffer == NULL)
+        return E_INVALIDARG;
+
+    // If we have GStreamer get buffer cllback set, then call it to get
+    // buffer. Otherwsie allocate memory internally.
+    if (GetGstBufferCallback != NULL && m_pGstBuffer == NULL)
+    {
+        GetGstBufferCallback(&m_pGstBuffer, (long)m_cbMaxLength, &m_CallbackData);
+        if (m_pGstBuffer == NULL)
+            return E_OUTOFMEMORY;
+
+        if (!gst_buffer_map(m_pGstBuffer, &m_GstMapInfo, GST_MAP_READWRITE))
+            return E_FAIL;
+
+        // Just in case check that we got right buffer size.
+        // GStreamer buffer can be bigger due to alligment.
+        if (m_GstMapInfo.maxsize < m_cbMaxLength)
+        {
+            gst_buffer_unmap(m_pGstBuffer, &m_GstMapInfo);
+            // INLINE - gst_buffer_unref()
+            gst_buffer_unref(m_pGstBuffer);
+            m_pGstBuffer = NULL;
+            return E_FAIL;
+        }
+
+        m_bUnmapGstBuffer = TRUE;
+
+        (*ppbBuffer) = m_GstMapInfo.data;
+    }
+    // Lock can be called multiple times, so if we have GStreamer buffer
+    // allocated just return it.
```

Comment:
Is it actually possible that Lock can be called multiple times? If it is, then I see that you don't keep track of the number of calls to Lock, so is it correct to assume that the caller does not match Lock / Unlock calls, but rather calls Unlock only once regardless of the number of calls to Lock?

Reply:
Yes, based on documentation Lock() can be called multiple times to acquire pointer and it should be valid until last Unlock() is called. The caller is responsible to match Lock / Unlock calls. So, current implementation is not correct. I will fixed it. Also, based on documentation Lock() can be called from separate threads, so it should be thread safe. I will fix it as well.